File: suffix_array.c
#include <string>
#include <iostream>
#define MAX_N = 100000

int n, k;
int rank[MAX_N + 1];
int tmp[MAX_N + 1];

bool compare_sa(int i, int j) {
	if (rank[i] != rank[j]) return rank[i] < rank[j];
	else {
		int ri = i + k <= n ? rank[i + k] : -1;
		int rj = j + k <= n ? rank[j + k] : -1;
		return ri < rj;
	}
}

void construct_sa(string S, int *sa) {
	n = S.length();
	for (int i = 0; i <= n; ++i) {
		sa[i] = i;
		rank[i] = i < n ? S[i] : -1;
	}
	for (int k = 1; k <= n; k = k << 1) {
		sort(sa, sa + n + 1, compare_sa);
		tmp[sa[0]] = 0;
		for (int i = 1; i <= n; ++i) {
			tmp[sa[p]] = tmp[sa[p - 1]] + (compare_sa(sa[i - 1], sa[i]) ? 1 : 0)
		}
		for (int i = 0; i <= n; ++i) {
			rank[i] = tmp[i];
		}
	}

}

void construct_lcp(string S, int *sa, int *lcp) {
	int n = S.length();
	for (int i = 0; i <= n; ++i) rank[sa[i]] = i;
	
	int h = 0;
	lcp[0] = 0;
	for (int i = 0; i < n; ++i) {
		int j = sa[rank[i] - 1];
		if (h > 0) h--;
		for (; j + h < n && i + h < n; h++) {
			if (S[j + h] != S[i + h]) break;
		}
	}
	lcp[rank[i] - 1] = h;
}

int main() {
	return 0;
}

File: suffix_array.py
import functools
class suffix_array:
    def __init__(self, s):
        self.s = s
        self.s_len = len(s)
        self.ranks = []
        self.lcps = []
        self.construct()

    def construct(self):
        sa = self.sa
        s_len = self.s_len
        ranks = self.ranks
        tmp = [ 0 for i in range(s_len + 1) ]
        for i in range(self.s_len + 1):
            sa.append(i)
            ranks.append(ord(self.s[i]) if i < n else -1)

        def compare_sa(i, j):
            if ranks[i] != ranks[j]:
                return ranks[i] < ranks[j]
            else:
                ri = ranks[i + k] if i + k <= self.s_len else -1
                rj = ranks[j + k] if j + k <= self.s_len else -1
                return ri < rj

        t = 1
        while t <= s_len:
            sorted(ranks, key = functools.cmp_to_key(compare_sa))
            tmp[sa[0]] = 0
            for p in range(1, s_len + 1):
                tmp[sa[p]] = tmp[sa[p - 1]] + 1 if compare_sa(sa[p - 1], sa[p]) else tmp[sa[p - 1]]
            for q in range(s_len + 1):
                ranks[q] = tmp[q]
            t *= 2
        self.sa = sa
        self.construct_lcp()

    def construct_lcp(self):
        h = 0
        self.lcp = [ 0 for i in range(self.s_len) ]
        for i in range(self.s_len):
            j = self.sa[self.ranks[i] - 1]
            if h > 0:
                h = h - 1
            while j + h < self.s_len and i + h < self.s_len:
                if self.s[j + h] != self.s[i + h]:
                    break
                h = h + 1
            self.lcps[self.ranks[i] - 1] = h

    def find(self, ts):
        l_ts = len(ts)
        l, h = 0, self.s_len
        while h - l > 1:
            m = (l + h) // 2
            if self.s[self.sa[m]:self.sa[m] + l_ts] < ts:
                l = m
            else:
                h = m
        if self.s[self.sa[h]:self.sa[h] + l_ts] == ts:
            return self.sa[h]
        return -1

sa = suffix_array('helloworld')
sa.construct()
for rank in sa.ranks:
    print('str: ', sa.s[:rank])
